### PR TITLE
[expected-type-api] Get the expected type for a given ST node position

### PR DESCRIPTION
**Pull request by DamithaSenevirathne**
Initial implementation of expected-type-api

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/expression_test.bal [hunk 1]
```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/expression_test.bal:3]:
```suggestion
// WSO2 LLC. licenses this file to you under the Apache License,
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/function_call_expression_test.bal [hunk 2]
```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/function_call_expression_test.bal:1]:
```suggestion
// Copyright (c) 2022 WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/function_call_expression_test.bal [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/function_call_expression_test.bal:3]:
```suggestion
// WSO2 LLC. licenses this file to you under the Apache License,
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/listener_declaration_test.bal [hunk 4]
```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/listener_declaration_test.bal:1]:
```suggestion
// Copyright (c) 2022 WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/listener_declaration_test.bal [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/listener_declaration_test.bal:3]:
```suggestion
// WSO2 LLC. licenses this file to you under the Apache License,
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/query_expression_test.bal [hunk 6]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/query_expression_test.bal:1]:
```suggestion
// Copyright (c) 2022 WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/query_expression_test.bal [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/query_expression_test.bal:3]:
```suggestion
// WSO2 LLC. licenses this file to you under the Apache License,
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/remote_action_node_test.bal [hunk 8]
```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/remote_action_node_test.bal:1]:
```suggestion
// Copyright (c) 2022 WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/remote_action_node_test.bal [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/remote_action_node_test.bal:3]:
```suggestion
// WSO2 LLC. licenses this file to you under the Apache License,
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/statement_test.bal [hunk 10]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/statement_test.bal:1]:
```suggestion
// Copyright (c) 2022 WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/statement_test.bal [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/statement_test.bal:3]:
```suggestion
// WSO2 LLC. licenses this file to you under the Apache License,
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/table_constructor_test.bal [hunk 12]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/table_constructor_test.bal:1]:
```suggestion
// Copyright (c) 2022 WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/table_constructor_test.bal [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/table_constructor_test.bal:3]:
```suggestion
// WSO2 LLC. licenses this file to you under the Apache License,
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/variable_declaration_tests.bal [hunk 14]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/variable_declaration_tests.bal:1]:
```suggestion
// Copyright (c) 2022 WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/variable_declaration_tests.bal [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/variable_declaration_tests.bal:3]:
```suggestion
// WSO2 LLC. licenses this file to you under the Apache License,
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/query_expression_test.bal [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+function testXMLType() {
+    xml data = xml `XML Data`;
+    from  in data
+}
+
+function testStringTypeExpression() {
+    string students = "Jane, Will";
+    from string name in ;
+}
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/query_expression_test.bal:42]: Missing new line

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/remote_action_node_test.bal [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+    remote function getVersion() returns string {
+        return "v1.0.0";
+    }
+}
+
+public function main() returns error? {
+    ProductClient prodClient = check new("http://example,com");
+    string v = prodClient->listAll();
+}
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expected-type/remote_action_node_test.bal:54]: missing new line at the end

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/ExpectedTypeFinder.java [hunk 18]
```diff
@@ -0,0 +1,1118 @@
...
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    public static String getAlias(QualifiedNameReferenceNode qNameRef) {
+        String alias = qNameRef.modulePrefix().text();
+        return getAlias(alias);
+    }
```
Comment (dulajdilshan) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/ExpectedTypeFinder.java:1047]: These methods are public, Is it intentional?

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/ExpectedTypeFinder.java [hunk 19]
```diff
@@ -0,0 +1,1118 @@
...
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static io.ballerina.compiler.api.impl.PositionUtil.isWithinParenthesis;
+import static io.ballerina.compiler.api.impl.PositionUtil.posWithinRange;
+import static io.ballerina.compiler.api.symbols.SymbolKind.MODULE;
+import static org.ballerinalang.model.tree.NodeKind.ERROR_CONSTRUCTOR_EXPRESSION;
+
+/**
+ * This transformer is used to resolve an expected type of given context.
```
Comment (dulajdilshan) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/ExpectedTypeFinder.java:136]:
```suggestion
 * This transformer is used to resolve an expected type for a given context.
```

**Review by dulajdilshan — COMMENTED** [review]
Added some minor suggestions for the first review phase

**Comment by DamithaSenevirathne** [comment]
Review comments are addressed.

**Review by dulajdilshan — APPROVED** [review]
LGTM

**Pull request merged**